codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketChunkedInput.java

```diff
@@ -0,0 +1,96 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http.websocketx;
+
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.websocketx.ContinuationWebSocketFrame;
+import io.netty.handler.codec.http.websocketx.WebSocketFrame;
+import io.netty.handler.stream.ChunkedInput;
+
+/**
+ * A {@link ChunkedInput} that fetches data chunk by chunk for use with WebSocket chunked transfers.
+ * <p>
+ * Each chunk from the input data will be wrapped within a {@link WebSocketFrame}. At the end of the input data,
+ * {@link WebSocketFrame} with finalFragment will be written.
+ * <p>
+ */
+public class WebSocketChunkedInput implements ChunkedInput<WebSocketFrame> {
+
+    private final ChunkedInput<ByteBuf> input;
+    private final ContinuationWebSocketFrame lastWebSocketContent;
+    private boolean sentLastChunk;
+    private final int rsv;
+
+    /**
+     * Creates a new instance using the specified input.
+     * @param input {@link ChunkedInput} containing data to write
+     */
+    public WebSocketChunkedInput(ChunkedInput<ByteBuf> input) {
+        this.input = input;
+        this.lastWebSocketContent = new ContinuationWebSocketFrame();
+        this.sentLastChunk = false;
+        this.rsv = 0;
+    }
+
+    /**
+     * Creates a new instance using the specified input.
+     * @param input {@link ChunkedInput} containing data to write
+     * @param lastWebSocketContent {@link ContinuationWebSocketFrame} that will be written as the terminating chunk.
+     *            Use this for training headers.
+     * @param rsv RSV1, RSV2, RSV3 used for extensions
+     */
+    public WebSocketChunkedInput(ChunkedInput<ByteBuf> input,
+                                 ContinuationWebSocketFrame lastWebSocketContent,
+                                 int rsv) {
+        this.input = input;
+        this.lastWebSocketContent = lastWebSocketContent;
+        this.sentLastChunk = false;
+        this.rsv = rsv;
+    }
+
+    @Override
+    public boolean isEndOfInput() throws Exception {
+        if (input.isEndOfInput()) {
+            // Only end of input after last WebSocket frame chunk has been sent
+            return sentLastChunk;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        input.close();
+    }
+
+    @Override
+    public WebSocketFrame readChunk(ChannelHandlerContext ctx) throws Exception {
+        if (input.isEndOfInput()) {
+            if (sentLastChunk) {
+                return null;
+            } else {
+                // Send last chunk for this input
+                sentLastChunk = true;
+                return lastWebSocketContent;
+            }
+        } else {
+            ByteBuf buf = input.readChunk(ctx);
+            return new ContinuationWebSocketFrame(false, rsv, buf);
+        }
+    }
```

Comment:
The special last frame is required for end of transmission , Otherwise chunked transmission is not complete.